server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1311,12 +1288,6 @@ protected DeletionStrategy deletionStrategyForOperation(final Delete delete) thr
 
     protected final DeletionStrategy planDeletionAsNonPrimary(Delete delete) throws IOException {
         assert assertNonPrimaryOrigin(delete);
-        maxSeqNoOfNonAppendOnlyOperations.updateAndGet(curr -> Math.max(delete.seqNo(), curr));
-        assert maxSeqNoOfNonAppendOnlyOperations.get() >= delete.seqNo() : "max_seqno of non-append-only was not updated;" +
-            "max_seqno non-append-only [" + maxSeqNoOfNonAppendOnlyOperations.get() + "], seqno of delete [" + delete.seqNo() + "]";
-        // unlike the primary, replicas don't really care to about found status of documents
-        // this allows to ignore the case where a document was found in the live version maps in
-        // a delete state and return true for the found flag in favor of code simplicity
         final DeletionStrategy plan;
         if (delete.seqNo() <= localCheckpointTracker.getProcessedCheckpoint()) {
```

Comment:
should we use `hasBeenProcessedBefore` here as well? And do the same for NOOPs?

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/43616/commits/b855a3a82260785ac83523d84c30a78244c4611b